Add exception on failed port string parsing in BindingAddress.cs Resolves #24384

h5aaimtron opened the pull request:
# Add exception on failed port string parsing in BindingAddress.cs

This PR adds a FormatException in the BindingAddress.cs when an invalid portString cannot be parsed as int.

## Description

Users have reported that when passing address strings containing invalid port strings is passed into BindingAddress.Parse(), instead of throwing an exception or generating a warning, the invalid port string is ignored, and the default ports (80/443) are applied. Invalid address strings such as "https://*:{abc}" should throw a FormatException due to the invalid port value 'abc' similar to how FormatExceptions are thrown for invalid hostnames.


Fixes #24384

h5aaimtron committed:
Implemented fix to match HubConnection.cs logic (https://github.com/dotnet/aspnetcore/issues/42806)

h5aaimtron committed:
Merge branch 'dotnet:main' into main

h5aaimtron committed:
Merge branch 'dotnet:main' into main

h5aaimtron committed:
Added port binding exception on invalid port string. Resolves #24384

h5aaimtron committed:
added tests for invalid port

gfoidl commented on src/Http/Http/src/BindingAddress.cs:
```diff
@@ -217,6 +217,8 @@ public static BindingAddress Parse(string address)
                     hasSpecifiedPort = true;
                     host = address.Substring(schemeDelimiterEnd, portDelimiterStart - schemeDelimiterEnd);
                     port = portNumber;
+                }else{
+                    throw new FormatException($"Invalid port: '{portString}'");
                 }
```
```suggestion
                }
                else
                {
                    throw new FormatException($"Invalid port: '{portString}'");
                }
```

gfoidl commented on src/Http/Http/src/BindingAddress.cs:
```diff
@@ -217,6 +217,8 @@ public static BindingAddress Parse(string address)
                     host = address.Substring(schemeDelimiterEnd, portDelimiterStart - schemeDelimiterEnd);
                     port = portNumber;
+                }else{
+                    throw new FormatException($"Invalid port: '{portString}'");
```
Is `FormatException` correct? In https://github.com/dotnet/aspnetcore/issues/24384#issuecomment-1261248502 it's just "more specific".

Maybe there should also be a "help-message" like

```suggestion
                    throw new FormatException($"Invalid port: '{portString}, only numbers are allowed'");
```
Although that should be obvious for ports...

h5aaimtron committed:
Update src/Http/Http/src/BindingAddress.cs

Applying suggestion to match linting.

Co-authored-by: Günther Foidl <gue@korporal.at>

h5aaimtron committed:
Update src/Http/Http/src/BindingAddress.cs

Adding additional help message.

Co-authored-by: Günther Foidl <gue@korporal.at>

h5aaimtron commented:
@gfoidl I added some tests in the PR as well.

gfoidl commented on src/Http/Http/src/BindingAddress.cs:
```diff
@@ -218,6 +218,10 @@ public static BindingAddress Parse(string address)
                 }
+                else
+                {
+                    throw new FormatException($"Invalid port: '{portString}, only numbers are allowed'");
```
```suggestion
                    throw new FormatException($"Invalid port: '{portString}', only numbers are allowed'");
```

h5aaimtron committed:
Update src/Http/Http/src/BindingAddress.cs

I think there's still an extra ' at the end after allowed. Perhaps a . or nothing to end the message?

Co-authored-by: Günther Foidl <gue@korporal.at>

h5aaimtron committed:
adjusted message

h5aaimtron commented:
Additional changes will be required to handle NamedPipes. It is odd that the previous code worked as it set a port with named pipes, so will require further research.

h5aaimtron committed:
Should clear the named pipe issue

h5aaimtron commented:
While reviewing the test failures, I identified a pre-existing issue in the code. The port parsing code is wrapped in an if statement checking it is not a unix pipe, but fails to check if it is not a named pipe as well. This allowed named pipes to enter the port parsing code block, which combined with my earlier change, resulted in the thrown FormatException. Since named pipes do not have ports and therefore do not need to parse port numbers, I've added the not named pipe condition to the if statement to skip the code block for named or unix pipes.

h5aaimtron committed:
removed invalid tests

h5aaimtron committed:
Remove port parse for IPv6 loop back address.

h5aaimtron committed:
adjust for IPv6

h5aaimtron commented:
Latest change fixes support for IPv6 [::1] binding. Test cases had non-bindable addresses marked as valid, so those were removed. For people looking simply to parse a Uri, they should use Uri.TryCreate(), not BindingAddress.Parse() as it infers a valid, bindable address, not just a valid uri.

h5aaimtron committed:
Merge branch 'dotnet:main' into main

ladeak commented on src/Http/Http/src/BindingAddress.cs:
```diff
@@ -203,10 +203,11 @@ public static BindingAddress Parse(string address)
+            // Verify not a loopback uri.
             var portDelimiterStart = address.LastIndexOf(':', pathDelimiterStart - 1, pathDelimiterStart - schemeDelimiterEnd);
-            if (portDelimiterStart >= 0)
+            if (portDelimiterStart >= 0 && address.Substring(pathDelimiterStart - 1, 1) != "]")
```
You could index into the address and compare against `]` to avoid the allocation.

`pathDelimiterStart - 1` is calculated already above.

Youssef1313 commented:
@danroth27 As 10.0.0 already shipped, I removed the milestone. If this really needs to get into 10.x, consider rebasing on the release/10.0 and change the PR target to release/10.0, it will then get added to the correct milestone [automatically](https://github.com/dotnet/aspnetcore/blob/8a687407a378a981dc1d7a4694d42ec8a735bdd6/.github/policies/resourceManagement.yml#L626-L636) (at the time of writing this comment, 10.0.8).
